cmd2/cmd2.py

```diff
@@ -2176,120 +2239,352 @@ def _cmdloop(self) -> bool:
+            self.perror(errmsg, traceback_war=False)
+            return
 
-        # The user is looking up an alias
-        if len(alias_arg_list) == 1:
-            if name in self.aliases:
-                self.poutput("alias {} {}".format(name, self.aliases[name]))
-            else:
-                self.perror("Alias {!r} not found".format(name), traceback_war=False)
+        if macro_name in self.aliases:
+            errmsg = "Macros cannot have the same name as an alias"
+            self.perror(errmsg, traceback_war=False)
+            return
 
-        # The user is creating an alias
-        else:
-            # Unquote redirection and pipes
-            index = 1
-            while index < len(alias_arg_list):
-                unquoted_arg = utils.strip_quotes(alias_arg_list[index])
-                if unquoted_arg in constants.REDIRECTION_TOKENS:
-                    alias_arg_list[index] = unquoted_arg
-                index += 1
-
-            # Build the alias value string
-            value = ' '.join(alias_arg_list[1:])
-
-            # Validate the alias to ensure it doesn't include weird characters
-            # like terminators, output redirection, or whitespace
-            valid, invalidchars = self.statement_parser.is_valid_command(name)
-            if valid:
-                # Set the alias
-                self.aliases[name] = value
-                self.poutput("Alias {!r} created".format(name))
-            else:
-                errmsg = "Aliases can not contain: {}".format(invalidchars)
-                self.perror(errmsg, traceback_war=False)
+        utils.unquote_redirection_tokens(args.command_args)
 
-    def complete_alias(self, text: str, line: str, begidx: int, endidx: int) -> List[str]:
-        """ Tab completion for alias """
-        alias_names = set(self.aliases.keys())
-        visible_commands = set(self.get_visible_commands())
+        # Build the macro value string
+        value = args.command
+        for cur_arg in args.command_args:
+            value += ' ' + cur_arg
 
-        index_dict = \
-            {
-                1: alias_names,
-                2: list(alias_names | visible_commands)
-            }
-        return self.index_based_complete(text, line, begidx, endidx, index_dict, self.path_complete)
+        # Find all normal arguments
+        arg_list = []
+        normal_matches = re.finditer(MacroArg.macro_normal_arg_pattern, value)
+        max_arg_num = 0
+        num_set = set()
 
-    @with_argument_list
-    def do_unalias(self, arglist: List[str]) -> None:
-        """Unsets aliases
+        while True:
+            try:
+                cur_match = normal_matches.__next__()
 
-    Usage: unalias [-a] name [name ...]
-    Where:
-        name - name of the alias being unset
+                # Get the number between the braces
+                cur_num = int(re.findall(MacroArg.digit_pattern, cur_match.group())[0])
```

Comment:
A better name would be `arg_num`, in my opinion, although your disdain for variable names that differ by a single letter would then be violated if you took my previous suggestion (so I'd suggest adding `ber` to one of them (as in `number`, not bit error rate))